src/backend/distributed/planner/fast_path_router_planner.c

```diff
@@ -0,0 +1,431 @@
+
+	/* there is only a single relation rte */
+	seqScanNode->scanrelid = 1;
+
+	plan->targetlist = copyObject(parse->targetList);
+	plan->qual = NULL;
+	plan->lefttree = NULL;
+	plan->righttree = NULL;
+	plan->plan_node_id = 1;
+
+	/*  rtable is used for access permission checks */
+	result->commandType = parse->commandType;
+	result->queryId = parse->queryId;
+	result->stmt_len = parse->stmt_len;
+
+	result->rtable = copyObject(parse->rtable);
+	result->planTree = (Plan *) plan;
+
+	relationId = ExtractFirstDistributedTableId(parse);
+	result->relationOids = list_make1_oid(relationId);
+
+	return result;
+}
+
+
+/*
+ * FastPathRouterQuery gets a query and returns true if the query is eligable for
+ * being a fast path router query.
+ * The requirements for the fast path query can be listed below:
+ *
+ *   - SELECT query without CTES, sublinks-subqueries, set operations
+ *   - The query should touch only a single hash distributed or reference table
+ *   - The distribution with equality operator should be in the WHERE clause
+ *      and it should be ANDed with any other filters. Also, the distribution
+ *      key should only exists once in the WHERE clause. So basically,
+ *          SELECT ... FROM dist_table WHERE dist_key = X
+ *   - No returning for UPDATE/DELETE queries
+ */
+bool
+FastPathRouterQuery(Query *query)
+{
+	RangeTblEntry *rangeTableEntry = NULL;
+	FromExpr *joinTree = query->jointree;
+	Node *quals = NULL;
+	Oid distributedTableId = InvalidOid;
+	Var *distributionKey = NULL;
+	DistTableCacheEntry *cacheEntry = NULL;
+
+	if (!EnableFastPathRouterPlanner)
+	{
+		return false;
+	}
+
+	if (!(query->commandType == CMD_SELECT || query->commandType == CMD_UPDATE ||
+		  query->commandType == CMD_DELETE))
+	{
+		return false;
+	}
+
+	/*
+	 * We want to deal with only very simple select queries. Some of the
+	 * checks might be too restrictive, still we prefer this way.
+	 */
+	if (query->cteList != NIL || query->returningList != NIL ||
+		query->hasSubLinks || query->setOperations != NULL ||
+		query->hasTargetSRFs || query->hasModifyingCTE)
+	{
+		return false;
+	}
+
+	/* make sure that the only range table in FROM clause */
+	if (list_length(query->rtable) != 1)
+	{
+		return false;
+	}
+
+	rangeTableEntry = (RangeTblEntry *) linitial(query->rtable);
+	if (rangeTableEntry->rtekind != RTE_RELATION)
+	{
+		return false;
+	}
+
+	/* we don't want to deal with append/range distributed tables */
+	distributedTableId = rangeTableEntry->relid;
+	cacheEntry = DistributedTableCacheEntry(distributedTableId);
+	if (!(cacheEntry->partitionMethod == DISTRIBUTE_BY_HASH ||
+		  cacheEntry->partitionMethod == DISTRIBUTE_BY_NONE))
+	{
+		return false;
+	}
+
+	/*
+	 * hasForUpdate is tricky because Citus does support only when
+	 * replication = 1 or reference tables.
+	 */
+	if (query->hasForUpdate)
+	{
+		if (cacheEntry->partitionMethod == DISTRIBUTE_BY_NONE ||
+			SingleReplicatedTable(distributedTableId))
+		{
+			return true;
```

Comment:
shouldn't we do the checks below?

Reply:
Ops, definitely this is wrong. I'll fix it.